### PR TITLE
Move version handling code out of PublishAction and into ChangeManager.

### PR DESCRIPTION
**Pull request by qz2017**
Refactor PublishAction class to only deal with publishing process and leave the change handling logic to ChangeManager.

### CODE REVIEW

File: apps/rush/src/utilities/ChangeManager.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+} from '@microsoft/rush-lib';
+
+import PublishUtilities, {
+  IChangeInfoHash
+} from './PublishUtilities';
+import ChangeFiles from './ChangeFiles';
+import PrereleaseToken from './PrereleaseToken';
+import ChangelogGenerator from './ChangelogGenerator';
+
+export default class ChangeManager {
```
Comment (nickpape-msft) [inline apps/rush/src/utilities/ChangeManager.ts]: >ChangeManager [](start = 21, length = 13)

comment about what the class does?

File: apps/rush/src/utilities/ChangeManager.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    prereleaseToken: PrereleaseToken = new PrereleaseToken(),
+    includeCommitDetails: boolean = false
+  ): void {
+    this._allPackages = this._rushConfiguration.projectsByName;
+
+    this._prereleaseToken = prereleaseToken;
+
+    this._changeFiles = new ChangeFiles(changesPath);
+    this._allChanges = PublishUtilities.findChangeRequests(
+      this._allPackages,
```
Comment (nickpape-msft) [inline apps/rush/src/utilities/ChangeManager.ts]: >this._allPackages [](start = 6, length = 17)

does this really need to be its own variable if we already store rushConfiguration?

**Comment by nickpape-msft** [comment]
:shipit:

File: apps/rush/src/utilities/ChangeManager.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      this._changeFiles,
+      includeCommitDetails,
```
Comment (nickpape-msft) [inline apps/rush/src/utilities/ChangeManager.ts]: >includeCommitDetails [](start = 6, length = 20)

I'm not sure what this variable is

File: apps/rush/src/utilities/ChangeManager.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    return this._orderedChanges && this._orderedChanges.length > 0;
+  }
+
+  public get changes(): IChangeInfo[] {
+    return this._orderedChanges;
+  }
+
+  public get allPackages(): Map<string, RushConfigurationProject> {
+    return this._allPackages;
+  }
```
Comment (nickpape-msft) [inline apps/rush/src/utilities/ChangeManager.ts]: this probably doesn't need to be exposed, as it already exists on passed-in rush config

**Comment by nickpape-msft** [comment]
:shipit:

File: apps/rush/src/utilities/ChangeManager.ts (hunk 1 above)
Comment (pgonzal) [inline apps/rush/src/utilities/ChangeManager.ts]: +1 I have no idea what any of these functions are supposed to do.  Please add comments

---
In reply to: [123110268](https://github.com/Microsoft/web-build-tools/pull/244#discussion_r123110268) [](ancestors = 123110268)

**Comment by pgonzal** [comment]
Please add a meaningful PR description -- did you change any actual functionality?  Why did we need this PR? etc.

File: apps/rush/src/actions/PublishAction.ts [hunk 5]
```diff
@@ -153,34 +153,21 @@ export default class PublishAction extends CommandLineAction {
...
-      allPackages,
-      changeFiles,
-      this._addCommitDetails.value,
-      this._prereleaseToken);
-    const orderedChanges: IChangeInfo[] = PublishUtilities.sortChangeRequests(allChanges);
-
-    if (orderedChanges.length > 0) {
+    changeManager.load(changesPath,
+      this._prereleaseToken,
+      this._addCommitDetails.value);
```
Comment (pgonzal) [inline apps/rush/src/actions/PublishAction.ts]: Would anyone really call load() without these parameters?  If not, why should they be optional?  It increases our test surface and could introduce mistakes if people forget to provide one of the parameters.

File: apps/rush/src/utilities/ChangeManager.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+  public apply(shouldCommit: boolean): void {
```
Comment (pgonzal) [inline apps/rush/src/utilities/ChangeManager.ts]: >apply [](start = 9, length = 5)

Comment?

File: apps/rush/src/utilities/ChangeManager.ts (hunk 2 above)
Comment (qz2017) [inline apps/rush/src/utilities/ChangeManager.ts]: This is pure refactoring. Code changes are minimized to keep history clear. 

---
In reply to: [123111221](https://github.com/Microsoft/web-build-tools/pull/244#discussion_r123111221) [](ancestors = 123111221)

File: apps/rush/src/utilities/ChangeManager.ts (hunk 3 above)
Comment (qz2017) [inline apps/rush/src/utilities/ChangeManager.ts]: This is pure refactoring. Code changes are minimized to keep history clear.  Improvements come in next PR.

---
In reply to: [123111312](https://github.com/Microsoft/web-build-tools/pull/244#discussion_r123111312) [](ancestors = 123111312)

File: apps/rush/src/utilities/ChangeManager.ts (hunk 4 above)
Comment (qz2017) [inline apps/rush/src/utilities/ChangeManager.ts]: It does. This is pure refactoring. Code changes are minimized to keep history clear.  Improvements come in next PR.

---
In reply to: [123111378](https://github.com/Microsoft/web-build-tools/pull/244#discussion_r123111378) [](ancestors = 123111378)

File: apps/rush/src/actions/PublishAction.ts (hunk 5 above)
Comment (qz2017) [inline apps/rush/src/actions/PublishAction.ts]: Test code actually calls load with default parameters.

---
In reply to: [123116612](https://github.com/Microsoft/web-build-tools/pull/244#discussion_r123116612) [](ancestors = 123116612)

**Pull request merged**